Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import GRPCCore
+import NIOCore
+
+// TODO: this is all done from client's perspective for now - work on server-side later.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+struct GRPCStreamStateMachine {
+  struct Configuration {
+    let maximumPayloadSize: Int
+    let compressor: Zlib.Compressor?
+    let decompressor: Zlib.Decompressor?
+  }
+
+  private var state: State
+  
+  enum State {
+    case clientIdleServerIdle(Configuration)
+    case clientOpenServerIdle(ClientOpenState)
+    case clientOpenServerOpen(ClientOpenState)
+    case clientOpenServerClosed(ClientOpenState)
```

Comment:
One thing I find really helpful to avoid bugs in state machines is having a separate state struct for each case. That way you can't accidentally switch to the wrong case by accident, e.g.:

```swift
switch state {
  case .clientOpenServerIdle(var state):
    state.doSomething()
    self.state = .clientOpenServerOpen(state) // oops, was meant to go back to idle
  default:
    // ...
}
```

You can also encode valid transitions by only having `init`s from other states, it makes the state transitions very deliberate.